### ISSUE TITLE
Lifetime mismatch on Clone does not raise an error

### ISSUE THREAD

**Issue by GGist** [rust-lang/rust#32543]
When implementing ```Clone``` for a type with lifetime ```'a``` and returning the same type with lifetime ```'static```, no compile error is generated and attempting to utilize the ```Clone``` function as intended leads to an "unrelated" compile error.

Example Code:
-------------------
```
use std::borrow::{Cow};

struct ByteContainer<'a> {
    bytes: Cow<'a, [u8]>
}

impl<'a> Clone for ByteContainer<'a> {
    fn clone(&self) -> ByteContainer<'static> {
        ByteContainer{ bytes: Cow::Owned((*self.bytes).to_vec()) }
    }
}

fn produce_byte_container() -> ByteContainer<'static> {
    let bytes = [0u8; 1];
    let container = ByteContainer{ bytes: Cow::Borrowed(&bytes[..]) };
    
    container.clone()
}

fn main() { }
```

Expected Result:
---------------------
Compile error that the returned lifetime for clone does not match the lifetime of Self.

Actual Result:
-----------------
```error: `bytes` does not live long enough```

May be vaguely related to #29149.

**Comment by bluss** [comment]
Everything seems to be working as it should actually.

The `impl` doesn't have to define a function of the same type (in this case the method `fn clone(&self) -> Self`. The type checker verifies it's a *subtype* which is a type that is *as good* as the expected type (the subtype can always be used in place of the expected type).

So, returning a `BytesContainer<'static>` is not exactly the same as was expected, but this type can always be used where a `BytesContainer<'a>` is needed, for any `'a`. So the type checker is happy with your trait implementation.

When it comes to *calling* `.clone()`, you can not use the signature in the `impl`, it will type check it against the method definition of the trait. So you can not take advantage of the `'static` lifetime for this particular type there.

**Comment by arielb1** [comment]
The type of an impl method is allowed to be a subtype of the trait method.

**Comment by GGist** [comment]
Oh, ok that definitely makes sense as to why an error is not thrown. Thanks bluss for the explanation, I was thinking about it completely wrong.
